src/sql/magic.py

```diff
@@ -410,18 +418,57 @@ def interactive_execute_wrapper(**kwargs):
             else:
                 with_ = self._store.infer_dependencies(command.sql_original, args.save)
                 if with_:
-                    command.set_sql_with(with_)
-                    display.message(
-                        f"Generating CTE with stored snippets: {pretty_print(with_)}"
-                    )
+                    query_type = get_query_type(command.sql_original)
+
+                    if query_type != "SELECT":
+                        display.message_warning(
+                            f"Your query is using the following snippets: \
+{', '.join(with_)}. {IF_NOT_SELECT_MESSAGE} CTE generation is disabled"
+                        )
+                    else:
+                        command.set_sql_with(with_)
+                        display.message(
+                            f"Generating CTE with stored snippets: \
+{pretty_print(with_)}"
+                        )
                 else:
                     with_ = None
         else:
+            query_type = get_query_type(command.sql_original)
             if args.with_:
                 raise exceptions.UsageError(
                     "Cannot use --with with CTEs, remove --with and re-run the cell"
                 )
 
+            dependencies = self._store.infer_dependencies(
+                command.sql_original, args.save
+            )
+
+            # The following bit of code checks for dependencies within the parentheses
+            # We do not care about the name outside the parentheses being the same as
+            # another table / snippet here and only check if there are dependencies
+            # in the parentheses.
+            # This is for generating more relevant warnings.
+            parentheses_content = re.findall(
+                r"\((.*)\)", command.sql_original.replace("\n", " ")
+            )
+
+            dependency_in_CTE = []
+            for query in parentheses_content:
+                for dependency in dependencies:
+                    if dependency in query:
+                        dependency_in_CTE.append(dependency)
```

Comment:
please move this to a separate function so we can unit test it.

also, this doesn't seem like the best way to parse this. check how we're inferring dependencies e.g., detect `some_table` in `SELECT * FROM some_table` using sqlglot. that sounds like a better way to parse this

Reply:
I could use that method, but it would become difficult to find out which tables in the query are important to us here.

For example, if we have the following:

```
WITH outer_table AS (
    SELECT * FROM some_snippet
) SELECT * FROM outer_table
```

There are two tables here, `some_snippet` and `outer_table`.
We only want to know if `some_snippet` is a snippet or not. We do not care if `outer_table` is a snippet because we are binding it to the value of `SELECT * FROM some_snippet` in our query, so it has a value in the local scope regardless of any existing snippet in the global scope named `outer_table`. This makes it irrelevant to the error message, which is why I have checked for dependencies within each parentheses only.

Reply:
Made it into a function
[function link](https://github.com/ploomber/jupysql/pull/878/files#diff-527caf90dd08297f5253e826bd98fea0b6f4fbd44fe9f20472976f77868dac06R689-R706)
Added tests
[test link](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2090-R2114)